h2/src/main/org/h2/expression/Operation.java

```diff
@@ -296,7 +295,8 @@ public Expression optimize(Session session) {
                             dataType = Value.TIMESTAMP;
                             return this;
                         case Value.DATE:
-                        case Value.TIMESTAMP: {
+                        case Value.TIMESTAMP:
+                        case Value.TIMESTAMP_TZ: {
```

Comment:
here 
    DATE + TIMESTAMP_TZ 
and 
    TIMESTAMP + TIMESTAMP_TZ 
becomes possible.

which does not seem like a good combination to me.

Reply:
Why?
```
Feature not supported: "DATE + TIMESTAMP WITH TIME ZONE"; SQL statement:
SELECT TIMESTAMP WITH TIME ZONE '2010-01-01 0:00:00Z' + DATE '2010-01-01' [50100-197]
```
There is no `switch` label for date-time types above `TIME`. Both TIMESTAMP + DATE and TIMESTAMP WITH TIME ZONE + DATE are rejected.

Reply:
Only TIME can be combined with other date-time data types.

However, I don't know why TIME + TIMESTAMP was accepted. This combination is still accepted, I decided not to touch it.

Reply:
PostgreSQL accepts TIME + DATE, TIME + TIMESTAMP and TIME + TIMESTAMP WITH TIME ZONE, but rejects TIME + TIME. MySQL accepts everything, but results are not useful at all even for TIME + DATE addition. Oracle doesn't have `TIME` data type.

May be reject TIME + TIME addition? (Subtraction should be fine.) Or leave it as is?

Reply:
TIME can be viewed as either an absolute time in the day, or an offset/duration, so adding it to other time-like things make some kind of sense.
Adding two absolute points in time together makes less sense.

So I'd say leave the TIME stuff alone.

Where does DATE + TIMESTAMP_TZ get rejected? I see a switch here that accepts DATE on the LHS and then below it a switch that accepts TIMESTAMP_TZ on the RHS.

Reply:
The execution goes here:
```Java
            } else if (DataType.isDateTimeType(l) || DataType.isDateTimeType(r)) {
                switch (opType) {
                case PLUS:
```
Values are swapped if necessary:
```Java
                    if (r != Value.getHigherOrder(l, r)) {
                        // order left and right: INT < TIME < DATE < TIMESTAMP
                        swap();
                        int t = l;
                        l = r;
                        r = t;
                    }
```
Now `l` has `DATE` type (it is “smaller”) even if it was on the right side. There is no label for this data type in the following switch:
```Java
                    switch (l) {
                    case Value.INT: {
                        // -- skipped --
                    }
                    case Value.DECIMAL:
                    case Value.FLOAT:
                    case Value.DOUBLE: {
                        // -- skipped --
                    }
                    case Value.TIME:
                        // -- skipped --
                    }
                    break;
```
Handling of `PLUS` operation is completed here on the `break`, so execution goes here:
```Java
                throw DbException.getUnsupportedException(
                        DataType.getDataType(l).name + " " +
                        getOperationToken() + " " +
                        DataType.getDataType(r).name);
```

Only `TIME` and numeric data types can be added to another date-time value.

Reply:
Skipped lines have `return` operators, so their execution don't fall to this `throw`.

This method is very large, may be extract date-time related optimizations in a separate method to simplify reading?